python/mxnet/gluon/probability/block/stochastic_block.py

```diff
@@ -0,0 +1,127 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+# pylint: disable=abstract-method
+"""Stochastic block class."""
+__all__ = ['StochasticBlock', 'StochasticSequential']
+
+from functools import wraps
+from ...block import HybridBlock
+from ...utils import _indent
+
+
+class StochasticBlock(HybridBlock):
+    """`StochasticBlock` extends `HybridBlock` to support accumulating loss
+    in the forward phase, which is extremely useful in building Bayesian Neural Network,
+    where the loss function is composed of a classification loss and a KL loss.
+
+    """
+
+    def __init__(self, **kwargs):
+        super(StochasticBlock, self).__init__(**kwargs)
+        self._losses = []
+        self._losscache = []
+
+    def add_loss(self, loss):
+        self._losscache.append(loss)
+
+    @staticmethod
+    def collectLoss(func):
+        """To accumulate loss during the forward phase, one could first decorate
+        hybrid_forward with `StochasticBlock.collectLoss,
+        and then collect the loss tensor `x` by calling self.add_loss(x).
+        For example, in the following forward function,
+        we generate samples from a Gaussian parameterized by `loc` and `scale` and
+        accumulate the KL-divergence between it and its prior into the block's loss storage.:
+        @StochasticBlock.collectLoss
+        def hybrid_forward(self, F, loc, scale):
+            qz = mgp.Normal(loc, scale)
+            # prior
+            pz = mgp.Normal(F.np.zeros_like(loc), F.np.ones_like(scale))
+            self.add_loss(mgp.kl_divergence(qz, pz))
+            return qz.sample()
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            # Loss from hybrid_forward
+            func_out = func(self, *args, **kwargs)
+            collected_loss = self._losscache
+            self._losscache = []
+            return (func_out, collected_loss)
+
+        return inner
+
+    def __call__(self, *args, **kwargs):
+		# pylint: disable=arguments-differ
+        out = super().__call__(*args, **kwargs)
+        self._losses.extend(out[1])
+        return out[0]
```

Comment:
@StochasticBlock.collectLoss is only needed when the users want to compute per-block loss such as KL divergence. If such loss is not demanded, the users can simply ignore this decorator or use default HybridBlock.

Reply:
Thanks for clarifying @szhengac. As per the current implementation, if user ignore this decorator and don't add it when they don't need it, an error will be raised as the `__call__` function always expects two outputs. If the error is correct and expected (ie. adding  `@StochasticBlock.collectLoss` is mandatory), why not remove the decorator for simplicity and handle it's implementation by default (in the `__call__`)?

Reply:
If it is not needed, the `out[1]` will be an empty list, so there is no error incurred here. One reason that we use this decorator is that we also need to use it in `StochasticSequential` for collecting losses over multiple blocks. And, as `self.add_loss(kl)` is called inside `forward` or `hybrid_forward`, I think we cannot move the implementation details to `__call__`.

Reply:
`out[1]` will only be an empty list if the user specifies `@StochasticBlock.collectLoss`. If they don't specify the decorator, `out[1]` will not exist and there'd be an error. We should have a good error message or at least document that users must use `@StochasticBlock.collectLoss`

Reply:
Oh yes, you are right. We need to think about it. Thanks for pointing this out.

Reply:
@leezu 
Thanks for pointing this out, a error should be raised when users is attempting to call self.add_loss in a function not decorated by `collectLoss`. I will look into this issue.

Reply:
Thank you @xidulu. Also note that this issue is present when users don't call collectLoss. That's because the __call__ method here is always called when users use StochasticBlock

Reply:
@leezu 

I add two checks here: https://github.com/apache/incubator-mxnet/pull/18403/files#diff-85458cf5116b137da8148bf5b38bcfaeR74
https://github.com/apache/incubator-mxnet/pull/18403/files#diff-85458cf5116b137da8148bf5b38bcfaeR78

To make it clearer, I list several possible situations:

1. Users call add_loss inside functions decorated by CollectLoss, add_loss appends losses into _losscache, _losscache would then get cleared in CollectLoss, len(_losscache) becomes 0 when __call__ is invoked.

2. Users call add_loss without using CollectLoss, add_loss appends losses into _losscache, _losscache  still contains value when entering  __call__, in this case, a exception will be raised.

3. Users use CollectLoss without calling add_loss, self._losses = out[1] = []

4. Users use StochasticBlock without calling CollectLoss or add_loss, len(out) == 1, out[1] will not be accessed.

Reply:
@leezu 
Update: I made further changes here to avoid confusion. Now the users are forced to use to collectLoss decorator in all cases, otherwise an exception would be raised.